### PR TITLE
[Enhancement] Add conference date and location column for  OSF meeting

### PR DESCRIPTION
**Pull request by haoyuchen1992**
## Purpose
After discussing with community member, add conference date and location column to OSF meeting and delete the column that show if submission is accepted. 

Resolved [OSF-4693]

### Before Changes:
screenshot 2015-10-23 12 09 26

### After Changes:
screenshot 2015-10-23 12 09 02

1. Add the new data (from Community) and update both dev_popluate_conference.py and  popluate_conference.py.
2. Change model and add three fields - end_date, start_date and location.

## Side Effects
None

### CODE REVIEW

**Comment by haoyuchen1992** [comment]
This may has issue on storing date on Firefox and Safari -- Related PR -- https://github.com/CenterForOpenScience/osf.io/pull/4529

**Comment by samanehsan** [comment]
Please resolve the merge conflicts

File: website/conferences/views.py [hunk 1]
```diff
@@ -241,7 +241,9 @@ def conference_view(**kwargs):
             continue
         meetings.append({
             'name': conf.name,
-            'active': conf.active,
+            'location': conf.location,
+            'end_date': conf.end_date.strftime("%d. %b %Y") if conf.end_date else None,
```
Comment (samanehsan) [inline website/conferences/views.py]: I don't think there should be a period after the day in `end_date` or `start_date`

File: website/static/js/meetings.js [hunk 2]
```diff
@@ -44,12 +50,36 @@ function Meetings(data) {
...
+                    }
                 },
                 {
-                    data: 'active', // Data field name
-                    sortInclude: true,
-                    custom: function() { return item.data.active ? 'Yes' : 'No'; }
+                    data: 'location',
+                    sortInclude : true,
+                    custom : function() {
+                        console.log(item.data.location);
```
Comment (samanehsan) [inline website/static/js/meetings.js]: Please remove this `console.log statement`

File: website/conferences/views.py (hunk 1 above)
Comment (haoyuchen1992) [inline website/conferences/views.py]: What do you mean by that? Could you give me a small example?

File: website/static/js/meetings.js (hunk 2 above)
Comment (haoyuchen1992) [inline website/static/js/meetings.js]: @samanehsan Sorry about that, that's a very bad missing.

File: scripts/dev_populate_conferences.py [hunk 3]
```diff
@@ -87,6 +104,9 @@ def main():
             'conferences',
             'bitss.jpg',
         ),
+        'location': 'Berkeley, CA',
+        'start_date': 'Dec 11 2014',
+        'end_date': 'Dec 12 2104',
```
Comment (samanehsan) [inline scripts/dev_populate_conferences.py]: Please fix this typo (the year should be 2014).

File: website/static/js/meetings.js [hunk 4]
```diff
@@ -44,12 +50,36 @@ function Meetings(data) {
...
+                        if (item.data.start_date === null) {
+                            return item.data.end_date;
+                        }
+                        if (item.data.end_date === null) {
+                            return item.data.start_date;
+                        }
+                        if (item.data.end_date === item.data.start_date) {
+                            return item.data.end_date;
+                        }
+                        return item.data.start_date + ' ~ ' + item.data.end_date;
```
Comment (samanehsan) [inline website/static/js/meetings.js]: I think a `-` would look better than `~` here. The `~` makes me think that the dates are approximate for some reason.

File: scripts/dev_populate_conferences.py [hunk 5]
```diff
@@ -116,6 +142,9 @@ def main():
         'name': 'International Convention of Psychological Science 2015',
         'info_url': None,
         'logo_url': 'http://icps.psychologicalscience.org/wp-content/themes/deepblue/images/ICPS_Website-header_990px.jpg',
+        'location': 'Amsterdam, The Netherlands',
+        'start_date': 'Mar 12 2014',
+        'end_date': 'Mar 14 2015',
```
Comment (samanehsan) [inline scripts/dev_populate_conferences.py]: Is the end date supposed to be `Mar 14 2014`?

File: website/conferences/views.py (hunk 1 above)
Comment (samanehsan) [inline website/conferences/views.py]: The date format looks like this: `29. Aug 2015 `, but I don't think there should be a period after the day. Instead, I think it should look like this: `29 Aug 2015`

**Comment by samanehsan** [comment]
Pass complete @haoyuchen1992 :cat2:

File: website/conferences/views.py (hunk 1 above)
Comment (haoyuchen1992) [inline website/conferences/views.py]: @samanehsan After checking multiple conferences' home page, all of them are using the format like `Aug 29, 2015`. Let's do this one!

Comment (samanehsan) [inline website/conferences/views.py]: Sounds good to me @haoyuchen1992 :+1:

**Comment by whit537** [comment]
Patch looks straightforward on first read-through. Mind merging `default` out here again, @haoyuchen1992, to resolve the merge conflicts? I'll test it out locally next chance I get ...

File: scripts/dev_populate_conferences.py [hunk 6]
```diff
@@ -435,9 +632,22 @@ def main():
     },
 }
 
+def clear_up_conf():
```
Comment (whit537) [inline scripts/dev_populate_conferences.py]: How about `clear_conferences` to maintain parallelism with `populate_conferences`?

File: scripts/dev_populate_conferences.py [hunk 7]
```diff
@@ -435,9 +632,22 @@ def main():
...
+    print 'Clear all the existing conferences'
+    confs = Conference.find()
+    for conf in confs:
+        conf.remove()
+        conf.save()
```
Comment (whit537) [inline scripts/dev_populate_conferences.py]: Is there a reason not to use `Conference.objects.all().delete()` here?

Comment (whit537) [inline scripts/dev_populate_conferences.py]: Yes: `Conference` is a `modularodm` model, not a Django model. :-)

**Comment by whit537** [comment]
For my information: under what circumstances would we run `scripts/populate_conferences.py` (as opposed to `dev_populate_conferences`)? Can we delete it?

File: scripts/populate_conferences.py [hunk 8]
```diff
@@ -544,6 +684,14 @@ def populate_conferences():
                 admin_objs.append(user)
             except ModularOdmException:
                 raise RuntimeError('Username {0!r} is not registered.'.format(email))
+
+        # Convert string into datetime object
+        date_format = '%b %d %Y'
+        if attrs['end_date']:
+            attrs['end_date'] = datetime.strptime(attrs['end_date'], date_format)
```
Comment (whit537) [inline scripts/populate_conferences.py]: Nice. I haven't actually used `strptime` before (I usually just do plain old string parsing on dates). Good to know about! :)

File: website/conferences/model.py [hunk 9]
```diff
@@ -19,6 +18,9 @@ class Conference(StoredObject):
     name = fields.StringField(required=True)
     info_url = fields.StringField(required=False, default=None)
     logo_url = fields.StringField(required=False, default=None)
+    location = fields.StringField(required=False, default=None)
+    start_date = fields.DateTimeField(default=None)
+    end_date = fields.DateTimeField(default=None)
     active = fields.BooleanField(required=True)
```
Comment (whit537) [inline website/conferences/model.py]: Do we care to remove this field in the database? What's the requirement, I wonder?

File: website/static/js/meetings.js [hunk 10]
```diff
@@ -16,20 +16,26 @@ function Meetings(data) {
...
+                    width: '45%',
                     sortType : 'text',
                     sort : true
                 },
                 {
                      title: 'Submissions',
-                     width : '25%',
+                     width : '15%',
                      sortType : 'number',
                      sort : true
```
Comment (whit537) [inline website/static/js/meetings.js]: Are these inner two indented an extra space relative to the outer two on purpose?

**Comment by whit537** [comment]
I'm working on reviewing this TTW. I believe I've found the page in question at `/meetings/`. It's empty of meetings, which I suppose means I need to run `scripts/dev_populate_conferences.py`.

screen shot 2015-12-22 at 8 37 08 pm

**Comment by whit537** [comment]
I don't see an invoke task in `tasks.py` that seems to be for `dev_populate_conferences`. Let's see what happens when I run the script directly ...

**Comment by whit537** [comment]
```
[osf.io]$ python scripts/dev_populate_conferences.py 
Traceback (most recent call last):
  File "scripts/dev_populate_conferences.py", line 9, in <module>
    from website import settings
ImportError: No module named website
[osf.io]$
```

**Comment by whit537** [comment]
Why is `website` not importable? It's in `cwd` and has an `__init__.py`. Is `cwd` not on `sys.path` for some reason?

**Comment by whit537** [comment]
Indeed. What gives?

```python
(Pdb) l
  1     #!/usr/bin/env python
  2     # encoding: utf-8
  3     import pdb; pdb.set_trace()
  4  -> import os
  5     
  6     from modularodm import Q
  7     from modularodm.exceptions import ModularOdmException
  8     
  9     from website import settings
 10     from website.app import init_app
 11     from website.conferences.model import Conference
(Pdb) import sys
(Pdb) pp sys.path
['/Users/whit537/workbench/cos/osf.io/scripts',
 '/Users/whit537/workbench/cos/osf.io/env/src/dataverse',
 '/Users/whit537/lib/python',
 '/Users/whit537/workbench/cos/osf.io/env/lib/python27.zip',
 '/Users/whit537/workbench/cos/osf.io/env/lib/python2.7',
 '/Users/whit537/workbench/cos/osf.io/env/lib/python2.7/plat-darwin',
 '/Users/whit537/workbench/cos/osf.io/env/lib/python2.7/plat-mac',
 '/Users/whit537/workbench/cos/osf.io/env/lib/python2.7/plat-mac/lib-scriptpackages',
 '/Users/whit537/workbench/cos/osf.io/env/lib/python2.7/lib-tk',
 '/Users/whit537/workbench/cos/osf.io/env/lib/python2.7/lib-old',
 '/Users/whit537/workbench/cos/osf.io/env/lib/python2.7/lib-dynload',
 '/Users/whit537/lib/python2.7',
 '/Users/whit537/lib/python2.7/plat-darwin',
 '/Users/whit537/lib/python2.7/lib-tk',
 '/Users/whit537/lib/python2.7/plat-mac',
 '/Users/whit537/lib/python2.7/plat-mac/lib-scriptpackages',
 '/Users/whit537/workbench/cos/osf.io/env/lib/python2.7/site-packages']
(Pdb)
```

**Comment by whit537** [comment]
```
(Pdb) import os
(Pdb) os.getcwd()
'/Users/whit537/workbench/cos/osf.io'
(Pdb)
```

:flushed: 

Is there something weird about my setup?

**Comment by whit537** [comment]
```
[osf.io]$ python
Python 2.7.9 (default, Jan  4 2015, 20:46:16) 
[GCC 4.2.1 Compatible Apple LLVM 6.0 (clang-600.0.56)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import os
>>> os.getcwd()
'/Users/whit537/workbench/cos/osf.io'
>>> import sys
>>> for p in sys.path: print p
... 

/Users/whit537/workbench/cos/osf.io/env/src/dataverse
/Users/whit537/lib/python
/Users/whit537/workbench/cos/osf.io/env/lib/python27.zip
/Users/whit537/workbench/cos/osf.io/env/lib/python2.7
/Users/whit537/workbench/cos/osf.io/env/lib/python2.7/plat-darwin
/Users/whit537/workbench/cos/osf.io/env/lib/python2.7/plat-mac
/Users/whit537/workbench/cos/osf.io/env/lib/python2.7/plat-mac/lib-scriptpackages
/Users/whit537/workbench/cos/osf.io/env/lib/python2.7/lib-tk
/Users/whit537/workbench/cos/osf.io/env/lib/python2.7/lib-old
/Users/whit537/workbench/cos/osf.io/env/lib/python2.7/lib-dynload
/Users/whit537/lib/python2.7
/Users/whit537/lib/python2.7/plat-darwin
/Users/whit537/lib/python2.7/lib-tk
/Users/whit537/lib/python2.7/plat-mac
/Users/whit537/lib/python2.7/plat-mac/lib-scriptpackages
/Users/whit537/workbench/cos/osf.io/env/lib/python2.7/site-packages
>>>
```

**Comment by whit537** [comment]
```
>>> import website
>>> website.__file__
'website/__init__.py'
>>>
```

:flushed:

**Comment by whit537** [comment]
I am amazed that I can import `website` when `cwd` is not on `sys.path`. In the script case `scripts` is ending up on `sys.path`.

**Comment by whit537** [comment]
I don't find any docs acking for `dev_populate_conferences`.

**Comment by whit537** [comment]
```
[osf.io]$ python scripts/dev_populate_conferences.py 
[website.util.paths]  WARNING: Skipping load of "webpack-assets.json" in DEBUG_MODE.
Traceback (most recent call last):
  File "scripts/dev_populate_conferences.py", line 667, in <module>
    main()
  File "scripts/dev_populate_conferences.py", line 17, in main
    init_app(set_backends=True, routes=False)
  File "scripts/../website/app.py", line 119, in init_app
    init_addons(settings, routes)
  File "scripts/../website/app.py", line 44, in init_addons
    addon = init_addon(app, addon_name, routes=routes)
  File "scripts/../website/addons/base/__init__.py", line 940, in init_addon
    addon_module = importlib.import_module(import_path)
  File "/Users/whit537/lib/python2.7/importlib/__init__.py", line 37, in import_module
    __import__(name)
  File "scripts/../website/addons/box/__init__.py", line 3, in <module>
    from website.addons.box import model, routes, utils
  File "scripts/../website/addons/box/model.py", line 4, in <module>
    from box import CredentialsV2, BoxClient
ImportError: cannot import name CredentialsV2
[osf.io]$
```

```diff
diff --git a/scripts/dev_populate_conferences.py b/scripts/dev_populate_conferences.py
index 399a538..f7bbb49 100644
--- a/scripts/dev_populate_conferences.py
+++ b/scripts/dev_populate_conferences.py
@@ -1,8 +1,10 @@
 #!/usr/bin/env python
 # encoding: utf-8
-
+import sys
 import os
 
+sys.path.append(os.path.join(os.path.dirname(__file__), '..'))
+
 from modularodm import Q
 from modularodm.exceptions import ModularOdmException
```

**Comment by whit537** [comment]
Zorg. O.o

```
[osf.io]$ python
Python 2.7.9 (default, Jan  4 2015, 20:46:16) 
[GCC 4.2.1 Compatible Apple LLVM 6.0 (clang-600.0.56)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import box
>>> from box import CredentialsV2
>>>
```

**Comment by whit537** [comment]
```
[osf.io]$ python scripts/dev_populate_conferences.py 
[website.util.paths]  WARNING: Skipping load of "webpack-assets.json" in DEBUG_MODE.
> /Users/whit537/workbench/cos/osf.io/website/addons/box/model.py(5)<module>()
-> from box import CredentialsV2, BoxClient
(Pdb) import box
(Pdb) box.__file__
'/Users/whit537/workbench/cos/osf.io/scripts/box/__init__.py'
(Pdb)
```

:dizzy_face:

**Comment by whit537** [comment]
Okay!

```
[osf.io]$ python scripts/dev_populate_conferences.py 
[website.util.paths]  WARNING: Skipping load of "webpack-assets.json" in DEBUG_MODE.
[website.addons.github.settings]  WARNING: No local.py settings file found
[website.addons.osfstorage.settings]  WARNING: No local.py settings file found
[website.addons.wiki.settings]  WARNING: No local.py settings file found
[root]  DEBUG: Setting storage backends
[root]  INFO: Sentry disabled; Flask's debug mode enabled
Clear all the existing conferences
[osf.io]$
```

```
diff --git a/scripts/dev_populate_conferences.py b/scripts/dev_populate_conferences.py
index 399a538..c2fee05 100644
--- a/scripts/dev_populate_conferences.py
+++ b/scripts/dev_populate_conferences.py
@@ -1,8 +1,10 @@
 #!/usr/bin/env python
 # encoding: utf-8
-
+import sys
 import os
 
+sys.path[0] = os.path.join(os.path.dirname(__file__), '..')
+
 from modularodm import Q
 from modularodm.exceptions import ModularOdmException
```

**Comment by whit537** [comment]
Drat! Not there yet, where are those meetings?

screen shot 2015-12-22 at 9 07 11 pm

**Comment by whit537** [comment]
Yay! I get to use [Mongs](http://whit537.org/mongs/) again. :-)

Empty tho ... I need to run a migration or a schema load or something?

screen shot 2015-12-22 at 9 11 30 pm

**Comment by whit537** [comment]
Here we go ...

screen shot 2015-12-22 at 9 15 17 pm

**Comment by whit537** [comment]
The conferences look like they're there in the db.

screen shot 2015-12-22 at 9 16 11 pm

**Comment by whit537** [comment]
Alright, so where is the code that is serving up `/meetings/`?

**Comment by whit537** [comment]
/me discovers [`meeting_landing.mako`](https://github.com/CenterForOpenScience/osf.io/blob/4dafb6f6de05d28aea9d59ac66337a662ba6c7ff/website/templates/public/pages/meeting_landing.mako)

**Comment by whit537** [comment]
I can influence the world.

screen shot 2015-12-22 at 9 30 07 pm

**Comment by whit537** [comment]
http://localhost:5000/static/public/js/meetings-page.js is 404

**Comment by whit537** [comment]
Closer ...

screen shot 2015-12-22 at 9 39 11 pm

**Comment by whit537** [comment]
`meetings` is empty [in `conference_view`](https://github.com/CenterForOpenScience/osf.io/blob/79b414197b0ebf2fed6f963e57f55bc1f1efeb07/website/conferences/views.py#L239) ...

**Comment by whit537** [comment]
`Conference.find` seems to be working ...

```
(Pdb) len(Conference.find())
46
(Pdb) conf = Conference.find()[0]
(Pdb) conf.location
u'Boston, MA'
(Pdb) conf.start_date
datetime.datetime(2015, 4, 9, 0, 0)
(Pdb) conf.end_date
datetime.datetime(2015, 4, 11, 0, 0)
(Pdb)
```

**Comment by whit537** [comment]
Welp. I guess it does what it says on the tin. :-)

screen shot 2015-12-22 at 9 55 33 pm

Now how do I get some submissions?

**Comment by whit537** [comment]
Okay! Now I can start. :-)

screen shot 2015-12-22 at 9 58 10 pm

```
[osf.io]$ tail -n1 website/settings/local.py 
CONFERENCE_MIN_COUNT = 0
[osf.io]$
```

**Comment by whit537** [comment]
Alright, UI looks pretty good. :-)

> Improve the column sorting UI: remove the arrows and make clickable to sort up and down

Was this requirement from [OSF-4693] dropped at some point?

**Comment by whit537** [comment]
I don't have permissions on this repo such that I can edit the description, so I will make a punchlist here:

### Punchlist

 - [ ] rename `clear_up_conf` to `clear_conferences` (https://github.com/CenterForOpenScience/osf.io/pull/4505#discussion_r48315516)
 - [ ] explain why we shouldn't delete `scripts/populate_conferences.py` (https://github.com/CenterForOpenScience/osf.io/pull/4505#issuecomment-166772974)
 - [ ] document the correct way to invoke `scripts/dev_populate_conferences.py` (https://github.com/CenterForOpenScience/osf.io/pull/4505#issuecomment-166774620 ff.)
 - [ ] explain why we're not removing `active` in the database (https://github.com/CenterForOpenScience/osf.io/pull/4505#discussion_r48316101)
 - [ ] explain or remove extra indenting (https://github.com/CenterForOpenScience/osf.io/pull/4505#discussion_r48316326)
 - [ ] implement the column sorting changes (https://github.com/CenterForOpenScience/osf.io/pull/4505#issuecomment-166793667)
 - [ ] merge `develop` out to here, resolving conflicts (https://github.com/CenterForOpenScience/osf.io/pull/4505#issuecomment-165936022)

Okay, @haoyuchen1992! That's all I got for now.

cc: @sloria to review my review. ;-)

**Comment by whit537** [comment]
@haoyuchen1992 I just learned on OSF-4693 that you're no longer working at OSF. Sorry for the noise! I'll be careful not to mention you again in here if you want to unsubscribe. Best wishes! :-)

**Comment by Ghalko** [comment]
@whit537 taking over this ticket.

File: website/static/js/meetings.js [hunk 11]
```diff
@@ -44,12 +50,35 @@ function Meetings(data) {
...
+                        return m('span.text-center', item.data.count );
+                    }
                 },
                 {
-                    data: 'active', // Data field name
-                    sortInclude: true,
-                    custom: function() { return item.data.active ? 'Yes' : 'No'; }
+                    data: 'location',
+                    sortInclude : true,
+                    custom : function() {
```
Comment (brianjgeiger) [inline website/static/js/meetings.js]: Why isn't this just item.data.location?

File: scripts/dev_populate_conferences.py [hunk 12]
```diff
@@ -9,18 +9,20 @@
...
     init_app(set_backends=True, routes=False)
     populate_conferences()
 
-
 MEETING_DATA = {
     'spsp2014': {
         'name': 'Society for Personality and Social Psychology 2014',
         'info_url': None,
         'logo_url': None,
+        'location': 'Austin, TX',
```
Comment (brianjgeiger) [inline scripts/dev_populate_conferences.py]: This seems silly to have this much data duplicated between the populate conferences script and this one. Can we DRY this up?

**Comment by brianjgeiger** [comment]
Pass done. 🐧 Fix merge conflicts, DRY scripts, see if you really need a function call for that one item.

**Pull request closed without merging**